File: smo_test.nim
import std/[random, math, sequtils, stats, strformat, sugar, times]
import smo
import problem/classification
import kernel/[gaussian, cache]

when isMainModule:
  randomize(42)

  # define training set
  let n = 10000
  let nft = 5
  let x = collect:
    for i in 0..<n:
      collect:
        for j in 0..<nft:
          rand(1.0)

  # define parameters
  let gamma = 1.5
  let lmbda = 0.01
  let kernel = newGaussianKernel(x, gamma).cache(2000)

  let yr = collect:
    for xi in x:
      xi.foldl(a + sin(math.PI * b), 0.0)
  let ym = yr.mean()
  let y = collect:
    for yi in yr:
      if yi > ym: +1.0 else: -1.0
  var p = newProblem(kernel, y, lmbda)
  p.maxAsum = 0.01 * n.float64
  p.smoothingParam = 4.0 * ln(2.0)
  p.shift = 0.0
  let res = smo(p, verbose=1000, shrinkingPeriod=n, logObjective=true)
  echo fmt"It took {res.steps} steps in {res.time:.1f} seconds..."
  echo kernel.cacheSummary()
  var
    asum = 0.0
    acnt = 0
  for ai in res.a:
    asum += ai.abs()
    if ai.abs() > 1e-6:
      acnt += 1

  echo "asum = ", asum
  echo "acnt = ", acnt
  echo "c = ", res.c
  # echo "eps = ", p.epsilon + res.c

File: smo_test.py
import time
import nimporter
import nimsmo
import numpy as np

np.random.seed(42)

n, nft = 500, 10
x = np.random.rand(n, nft)
yr = np.sin(np.pi * x).sum(axis=1)
y = np.where(yr > np.median(yr), 1.0, -1.0)

lmbda = 0.01
gamma = 1.5
smoothingParam = 0.02

t0 = time.time()
res = nimsmo.solve(x.tolist(), y.tolist(), lmbda, gamma, smoothingParam)
t1 = time.time()
print(t1 - t0)

a = np.array(res['a'])
b = res['b']
c = res['c']
idx_sv = a != 0
print(f'N(SV) = {idx_sv.sum()} = {idx_sv.mean() * 100:.1f}%')

def kernel(x1, x2, gamma):
  x1sqr = (x1 * x1).sum(axis=1)
  x2sqr = (x2 * x2).sum(axis=1)
  return np.exp(-gamma * (x1sqr[:, None] + x2sqr[None, :] - 2.0 * x1.dot(x2.T)))


def smooth_max_p2(t, s):
  return np.piecewise(t, [t <= -s, t >= s], [0.0, lambda t: t, lambda t: 0.25 / s * (t + s) ** 2])

def dsmooth_max_p2(t, s):
  return np.piecewise(t, [t <= -s, t >= s], [0.0, 1.0, lambda t: 0.5 / s * (t + s)])

def ddsmooth_max_p2(t, s):
  return np.piecewise(t, [t <= -s, t >= s], [0.0, 0.0, 0.5 / s])

x_sv = x[idx_sv]
k = kernel(x, x_sv, gamma)

ka = k.dot(a[idx_sv]) / lmbda
d = ka + b

val = smooth_max_p2(1.0 - y * d, smoothingParam)
obj = 0.5 * ka.dot(a) + val.sum()
print(obj)

print(a - y * dsmooth_max_p2(1.0 - y * d, smoothingParam))

import torch